### PR TITLE
ObjectSeeker Defense Implementation

### LINKED ISSUE #2193: Implementation of ObjectSeeker Certifiably Robust Defense
**Is your feature request related to a problem? Please describe.**
ObjectSeeker is a certifiably robust defense against patch attacks on object detection models. The defense itself was originally intended for evasion patches, but may be generalizable for poisoning since it is patch agnostic. Therefore, this defense may be a good baseline against the BadDet poisoning attack on object detection models.

Paper link: <https://arxiv.org/abs/2202.01811>

**Describe the solution you'd like**
Since this is a certifiably robust defense, this should be implemented under `art.estimators.certification`. Just like the `randomized_smoothing` submodule, an `object_seeker` submodule will be created which will have the PyTorch implementation for now (TensorFlow may be added later).

The `ObjectSeekerPyTorch` class will take in an object detection model (Faster R-CNN or YOLO) and implement the corresponding `fit`, `predict`, and `certify` methods that all ART certification estimators typically do.

**Describe alternatives you've considered**
This may also be implemented somewhere under `art.defenses` since this is a defense, but it makes more sense under `art.estimators.certification` since it is a certifiably robust defense.

**Additional context**
A PyTorch implementation will only be done for now since object detection models (Faster R-CNN and YOLO) are typically only used in PyTorch. A TensorFlow implementation may be done in the future, but this is low priority.

### PR DESCRIPTION
**Pull request by f4str**
# Description

Implement the ObjectSeeker certifiably robust defense for object detectors. This is an estimator located in `art.estimators.certification.object_seeker` as the `PyTorchObjectSeeker` class. It is constructed in a similar way as the `PyTorchObjectDetector` or `PyTorchYolo`. Currently, there is only a PyTorch implementation, but a generic and TensorFlow implementation is possible in the future.

Additionally created utility functions for `intersection_over_union`, `intersection_over_area`, and `non_maximum_supression` under `art.utils`.

Note that unit tests are in `tests/estimators/object_detection` rather than `tests/estimators/certification` since a YOLO model is required.

Relevant links:
* Paper: <https://arxiv.org/abs/2202.01811>
* Code: <https://github.com/inspire-group/ObjectSeeker>

Fixes #2193 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for ObjectSeeker YOLO
- [x] Unit tests for ObjectSeeker Faster R-CNN

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/estimators/object_detection/test_object_seeker.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+import pytest
+
+from art.estimators.certification.object_seeker import PyTorchObjectSeeker
+from tests.utils import ARTTestException
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture()
+def get_pytorch_yolo(get_default_cifar10_subset):
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_object_seeker.py:32]: Is this a fixture that we also could use in other tests and move to contest.py?

File: tests/estimators/object_detection/test_object_seeker.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        # Compute loss before training
+        loss1 = object_seeker.compute_loss(x=x_test, y=y_test)
+
+        # Train for one epoch
+        object_seeker.fit(x_test, y_test, nb_epochs=1)
+
+        # Compute loss after training
+        loss2 = object_seeker.compute_loss(x=x_test, y=y_test)
+
+        assert loss1 != loss2
```
Comment (beat-buesser) [inline tests/estimators/object_detection/test_object_seeker.py:128]: Should `loss2` be smaller than `loss1`?

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for your well prepared pull request implementing ObjectSeeker in ART! I have only a few questions and suggestions, but otherwise the code looks good.

File: art/estimators/certification/object_seeker/object_seeker.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+from art.utils import intersection_over_area, non_maximum_suppression
+
+logger = logging.getLogger(__name__)
+
+
+class ObjectSeekerMixin(abc.ABC):
+    """
+    Implementation of the ObjectSeeker certifiable robust defense applied to object detection models.
+    The original implementation is https://github.com/inspire-group/ObjectSeeker
```
[the remark this answers is not in the document]
Comment (f4str) [reply art/estimators/certification/object_seeker/object_seeker.py:41]: The original code has an MIT license, so we can probably include their license below the ART one.

File: tests/estimators/object_detection/test_object_seeker.py (hunk 2 above)
Comment (f4str) [reply tests/estimators/object_detection/test_object_seeker.py:128]: I borrowed the same unit test from `test_pytorch_yolo.py`. Logically it should be less than, but I believe it was not equal in case of stochasticity that causes the loss to increase. I can look into this and see if it can be changed to less than for both tests.

**Comment by f4str** [comment]
Hi @beat-buesser thank you for the review. I've addressed the changes by adding the license from the original ObjectSeeker repo and extracting the fixtures into a `conftest.py`. I've extracted both the YOLO and Faster R-CNN fixtures and created a new unit test for Faster R-CNN. Now ObjectSeeker will test both object detector models.

As for the training loss not-equal vs less-than, it needs to remain a not-equal condition since due to stochasticity, the loss does not always decrease, especially with the "incorrect" dataset. The same applies to the PyTorchYolo and PyTorchFasterRCNN models.

**Comment by beat-buesser** [comment]
Hi @f4str Thank you very much for the updates as well as checking and documenting! We'll merge this PR as soon as possible.
